### PR TITLE
Deprecating `localai` in favor of expanded `openai_like`

### PR DESCRIPTION
**Pull request by jamesbraza**
# Description

- Deprecates `localai.LocalAI` in favor of `openai_like.OpenAILike` with special parameterization
- Expands `OpenAI._update_max_tokens` to have a backdoor for not adding `max_tokens`
- Avoids duplication of `OpenALike` `Field` descriptions
- Fixes testing in several ways
    - Stub tokenizer now delivers `List[int]`
    - Local testers can now run without `OPENAI_API_KEY` in their environment
- Tightens test assertions

## Type of Change

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

- [x] Added new unit/integration tests
- [ ] Added new notebook (that tests end-to-end)
- [x] I stared at the code and made sure it makes sense

### CODE REVIEW

File: llama_index/llms/openai_like.py [hunk 1]
```diff
@@ -58,10 +71,19 @@ def _tokenizer(self) -> Tokenizer:
                     "huggingface tokenizers with OpenAILike."
                 ) from exc
 
-            return AutoTokenizer.from_pretrained(self._get_model_name())
```
Comment (jamesbraza) [inline llama_index/llms/openai_like.py:61]: This wasn't quite correct, the model `OpenAILike` wraps here isn't a tokenizer. `_get_model_name` will 9/10 times return a non-tokenizer model.

We needed the `tokenizer` field to be flexible to this.

File: llama_index/llms/openai_like.py [hunk 2]
```diff
@@ -58,10 +71,19 @@ def _tokenizer(self) -> Tokenizer:
...
+            return AutoTokenizer.from_pretrained(self.tokenizer)
+        return self.tokenizer
 
     @classmethod
     def class_name(cls) -> str:
         return "OpenAILike"
+
+
+# Use these as kwargs for OpenAILike to connect to LocalAIs
+localai_defaults: Mapping[str, Any] = MappingProxyType(
```
Comment (jamesbraza) [inline llama_index/llms/openai_like.py:83]: Here is where `LocalAI` lives now, it's just a defaulted parameterization of `OpenAILike` 🚀

File: tests/llms/test_openai_like.py [hunk 3]
```diff
@@ -2,56 +2,98 @@
...
 
 from llama_index.llms import OpenAILike
-from llama_index.llms.base import ChatMessage
+from llama_index.llms.base import ChatMessage, MessageRole
+from llama_index.llms.openai import Tokenizer
+from llama_index.llms.openai_like import localai_defaults
 
 
-class MockTokenizer:
-    def encode(self, text: str) -> List[str]:
```
Comment (jamesbraza) [inline tests/llms/test_openai_like.py:9]: This type hint wasn't matching the `Tokenizer` type, subclassing `Tokenizer` exposes this.

Also, just fyi when naming these things, it's not a `MagicMock` or `Mock` subclass, so they should be named "stub."  Read here for more info: https://martinfowler.com/articles/mocksArentStubs.html

Cc @Disiok

File: llama_index/llms/openai.py [hunk 4]
```diff
@@ -68,10 +68,16 @@ class OpenAI(LLM):
     )
     max_retries: int = Field(description="The maximum number of API retries.")
 
-    api_key: str = Field(default=None, description="The OpenAI API key.", exclude=True)
```
Comment (jamesbraza) [inline llama_index/llms/openai.py:71]: Having `default=None` was a type mismatch, even though it gets updated in `__init__`

**Comment by jamesbraza** [comment]
Also cc @benbot, I can't directly add you as a PR reviewer it seems

File: llama_index/llms/localai.py [hunk 5]
```diff
@@ -1,75 +0,0 @@
...
-from llama_index.llms.openai import OpenAI
-from llama_index.llms.openai_utils import is_function_calling_model
-
-DEFAULT_API_KEY = "fake"
-DEFAULT_API_HOST = "localhost"
-DEFAULT_API_PORT = 8080
-DEFAULT_API_BASE = f"{DEFAULT_API_HOST}{DEFAULT_API_PORT}"
-
-
-class LocalAI(OpenAI):
```
Comment (jamesbraza) [inline llama_index/llms/localai.py:15]: This is now deprecated in favor of the more generalized `OpenAILike`, thanks @benbot for enabling this

**Comment by sfxworks** [comment]
Will there also be embedding support with this?

**Comment by jamesbraza** [comment]
@sfxworks it's a good question. Neither of the original `LocalAI` or `OpenAILike` here support embeddings.  This PR doesn't add embeddings support, it just merges `LocalAI` into a parameterization of `OpenAILike`.

Embeddings support for OpenAI-like models is a separate ask. There's definitely good possibility for either:

- Generalizing `OpenAIEmbedding` to work with an OpenAI-like embedder
- Creating a new `OpenAILikeEmbedding` class

File: llama_index/llms/openai_like.py (hunk 1 above)
Comment (nerdai) [reply llama_index/llms/openai_like.py:61]: Makes sense I think given the logic of _get_model_name checks for `ft` and splits the string, which I think is pretty narrow to OpenAI. What model names were you testing out of curiousity?

File: llama_index/llms/openai_like.py (hunk 2 above)
Comment (nerdai) [reply llama_index/llms/openai_like.py:83]: nice!

**Review by nerdai — APPROVED** [review]

**Comment by nerdai** [comment]
This looks pretty clean @jamesbraza! Thanks for doing this. It seems we got good ol' conflicts with `CHANGELOG.md`.

Before merging just want to check with @Disiok if there is anything special we need to do since this would deprecating `LocalAi`.

File: llama_index/llms/openai_like.py (hunk 1 above)
Comment (jamesbraza) [reply llama_index/llms/openai_like.py:61]: With `LocalAI` or `llama-cpp-python`, you get access to many models. Check LocalAI's model gallery to get some ideas: https://localai.io/models/ --> https://github.com/go-skynet/model-gallery

For example, their Hugging Face configs are pretty extensive: https://raw.githubusercontent.com/go-skynet/model-gallery/main/huggingface.yaml

File: tests/llms/test_openai_like.py (hunk 3 above)
Comment (Disiok) [reply tests/llms/test_openai_like.py:9]: Thanks for the fix!

File: llama_index/llms/openai_like.py [hunk 6]
```diff
@@ -22,34 +23,46 @@ class OpenAILike(OpenAI):
 
     context_window: int = Field(
         default=DEFAULT_CONTEXT_WINDOW,
-        description="The maximum number of context tokens for the model.",
+        description=LLMMetadata.__fields__["context_window"].field_info.description,
+    )
+    num_output: Optional[int] = Field(
```
Comment (Disiok) [inline llama_index/llms/openai_like.py:28]: why do we need `num_output` as well as `max_tokens`? Don't they represent the same thing?

File: llama_index/llms/openai_like.py (hunk 2 above)
Comment (Disiok) [inline llama_index/llms/openai_like.py:83]: should we maybe add a notebook to show how to do this?

File: CHANGELOG.md [hunk 7]
```diff
@@ -1,5 +1,11 @@
 # ChangeLog
 
+## Unreleased
+
+### New Features
+
+- Consolidates `LocalAI` into a parameterization of `OpenAILike` (#8241)
```
Comment (Disiok) [inline CHANGELOG.md:7]: could we mark this as breaking change?

File: llama_index/llms/__init__.py [hunk 8]
```diff
@@ -24,7 +24,6 @@
 from llama_index.llms.langchain import LangChainLLM
 from llama_index.llms.litellm import LiteLLM
 from llama_index.llms.llama_cpp import LlamaCPP
-from llama_index.llms.localai import LocalAI
```
Comment (Disiok) [inline llama_index/llms/__init__.py:27]: could we maybe just change LocalAI to be an alias for `OpenAILike` to have some compatibility from before?

File: llama_index/llms/openai_like.py (hunk 2 above)
Comment (jamesbraza) [reply llama_index/llms/openai_like.py:83]: Yeah I can make a Notebook, wanted to get team onboard before I invested the time in a demo Notebook

File: llama_index/llms/openai_like.py (hunk 6 above)
Comment (jamesbraza) [reply llama_index/llms/openai_like.py:28]: Haha nice catch, reading the added code here, it does seem like I added some indirection 😅 .

I will investigate this later when making the demo notebook, and likely conclude to remove this extra `Field` 👍

File: llama_index/llms/__init__.py (hunk 8 above)
Comment (jamesbraza) [reply llama_index/llms/__init__.py:27]: My thinking was cuz both `LocalAI` and `OpenAILike` are only like a month old, it's okay to remove `LocalAI`, with a good `CHANGELOG` entry.

Are you wanting `LocalAI` to just be in the `llms` namespace?  I am wondering if a good `docs/` article would be sufficient

Comment (sfxworks) [reply llama_index/llms/__init__.py:27]: FWIW I'm open to updating the docs I updated recently on LocalAI's end https://github.com/mudler/LocalAI/pull/1237

**Comment by jamesbraza** [comment]
Oh wow @sfxworks I didn't realize you're actually using `LocalAI` haha, awesome!  Now I will have us do a deprecation cycle for `LocalAI`, since it's actually being used.

This PR has been open for a few weeks, it's too bad we didn't move faster here, this change could've been merged before you made that demo. 

```python
llm = LocalAI(temperature=0, model_name="gpt-3.5-turbo", api_base="http://local-ai.default", api_key="stub")
llm.globally_use_chat_completions = True;
```

A few questions/comments on that part ^^^:
- http://local-ai.default: why did you select this URL for `api_base` over something like 127.0.0.1?
- The `LocalAI` class automatically passes `api_key="fake"`, no need to have `api_key="stub"` in the signature
- You can have `globally_use_chat_completions` be directly in the signature, it could be all one line like the below

```python
llm = LocalAI(
    temperature=0, model_name="gpt-3.5-turbo", api_base="http://local-ai.default",
    globally_use_chat_completions=True
)
```

**Comment by sfxworks** [comment]
> * http://local-ai.default: why did you select this URL for `api_base` over something like 127.0.0.1?
I'm running localai "locally" on lan, specifically in a k8s deployment. This example runs in a python container in the cluster and uses the k8s dns naming convention to reference the deployment. 

```
NAME               TYPE           CLUSTER-IP      EXTERNAL-IP    PORT(S)        AGE
service/local-ai   LoadBalancer   10.96.171.118   172.16.1.214   80:30364/TCP   31d

NAME                       READY   UP-TO-DATE   AVAILABLE   AGE
deployment.apps/local-ai   1/1     1            1           35d
```

> * The `LocalAI` class automatically passes `api_key="fake"`, no need to have `api_key="stub"` in the signature
> * You can have `globally_use_chat_completions` be directly in the signature, it could be all one line like the below
Ah good to know. Thanks!

**Comment by jamesbraza** [comment]
Dang there were a ton of change in the last month 😆 

I am breaking this PR up, instead of dealing with a ton of rebase conflicts. See: 
- Subset of tests to https://github.com/run-llama/llama_index/pull/9026

**Comment by jamesbraza** [comment]
This has now been completely decomposed and merged in across 3 PRs.  All changes are available in today's [0.9.9 release](https://github.com/run-llama/llama_index/releases/tag/v0.9.9).  Closing this out.

---

@sfxworks do you need help updating `LocalAI` docs? Feel free to tag me on a PR, or if you don't have the bandwidth I can update

```diff
- from llama_index.llms import LocalAI
- llm = LocalAI(temperature=0, model_name="gpt-3.5-turbo", api_base="http://local-ai.default", api_key="stub")
- llm.globally_use_chat_completions = True;
+ from llama_index.llms import LOCALAI_DEFAULTS, OpenAILike
+ llm = OpenAILike(temperature=0, model_name="gpt-3.5-turbo", is_chat_model=True, **LOCALAI_DEFAULTS)
```

**Pull request closed without merging**